### ISSUE TITLE
Custom FromParams need to also accept **extras

### ISSUE THREAD

**Issue by codedecde** [allenai/allennlp#2665]
For classes with custom from_params (eg: FeedForward), the from_params method needs to accept extras. Otherwise, if being instantiated from a class with extras (Eg: Model), the current FromParams implementation passes them to the class, which results it an error.

Steps to reproduce the behavior
Consider the following code snippet
```
@Model.register("tmp_model")`
class TmpModel(Model):
    def __init__(self, vocab: Vocabulary, feed_forward_list: List[FeedForward]):
        super(TmpModel, self).__init__(vocab=vocab)
        self.feed_forward = nn.ModuleList(feed_forward_list)

params = Params({
    "type": "tmp_model",
    "feed_forward": [{"input_dim": 512, "num_layers": 1, "hidden_dims": [2], "activations": ["linear"]}]
})

vocab = Vocabulary()
model = Model.from_params(vocab=vocab, params=params)
```

**Error**
from_params() got an unexpected keyword argument 'vocab'

**System:**
 - OS: [e.g. OSX, Linux]
 - AllenNLP version: [e.g. v0.8.2, or "I installed from master"]

**Comment by vidurj** [comment]
I am not sure I understand the issue. The `from_params` method does accept extras, and your example works with the `Vocabulary` being provided as an extra. Could you rephrase your question?

```
from allennlp.models import Model
from allennlp.modules import FeedForward
from allennlp.common.params import Params
from allennlp.data.vocabulary import Vocabulary

@Model.register('tmp_model')
class TmpModel(Model):
    def __init__(self, vocab: Vocabulary, feed_forward: FeedForward):
        super(TmpModel, self).__init__(vocab=vocab)
        self.feed_forward = feed_forward


params = Params({
        "type": "tmp_model",
        "feed_forward": {"input_dim": 512, "num_layers": 1, "hidden_dims": [2], "activations": ["linear"]}
})

vocab = Vocabulary()
model = Model.from_params(vocab=vocab, params=params)
```

**Comment by kl2806** [comment]
Closing due to inactivity, feel free to reopen if you still have an issue.

**Comment by codedecde** [comment]
Sorry for the delayed response.
So, I think the issue is with 
https://github.com/allenai/allennlp/blob/deae1db203efa9400b1d8b479b4faca3876a907a/allennlp/common/from_params.py#L204
So, if a parameter is a List[FeedForward], then that would cause an issue, since FeedForward does not accept extras.

**Comment by matt-gardner** [comment]
Hmm, I took a quick look at the logic, but the bug wasn't immediately apparent to me.  Probably @joelgrus is the right person to look at this, when you have time.

**Comment by codedecde** [comment]
One potential fix could be defining
```python
def create_extras(cls: Type[T], extras: Dict[str, Any]) -> Dict[str, Any]:
    subextras: Dict[str, Any] = {}
    if takes_arg(cls.from_params, "extras"):
        # If annotation.params accepts **extras, we need to pass them all along.
        # For example, `BasicTextFieldEmbedder.from_params` requires a Vocabulary
        # object, but `TextFieldEmbedder.from_params` does not.
        subextras = extras
    else:
        # Otherwise, only supply the ones that are actual args; any additional ones
        # will cause a TypeError.
        subextras = {k: v for k, v in extras.items()
                     if takes_arg(cls.from_params, k)}
    return subextras
```
And then using it in create_kwargs as 
```python
def create_kwargs(cls: Type[T], params: Params, **extras) -> Dict[str, Any]:
    """
    Given some class, a `Params` object, and potentially other keyword arguments,
    create a dict of keyword args suitable for passing to the class's constructor.

    The function does this by finding the class's constructor, matching the constructor
    arguments to entries in the `params` object, and instantiating values for the parameters
    using the type annotation and possibly a from_params method.

    Any values that are provided in the `extras` will just be used as is.
    For instance, you might provide an existing `Vocabulary` this way.
    """
    # Get the signature of the constructor.
    from allennlp.models.archival import load_archive  # import here to avoid circular imports

    signature = inspect.signature(cls.__init__)
    kwargs: Dict[str, Any] = {}

    # Iterate over all the constructor parameters and their annotations.
    for name, param in signature.parameters.items():
        # Skip "self". You're not *required* to call the first parameter "self",
        # so in theory this logic is fragile, but if you don't call the self parameter
        # "self" you kind of deserve what happens.
        if name == "self":
            continue

        # If the annotation is a compound type like typing.Dict[str, int],
        # it will have an __origin__ field indicating `typing.Dict`
        # and an __args__ field indicating `(str, int)`. We capture both.
        annotation = remove_optional(param.annotation)
        origin = getattr(annotation, '__origin__', None)
        args = getattr(annotation, '__args__', [])

        # The parameter is optional if its default value is not the "no default" sentinel.
        default = param.default
        optional = default != _NO_DEFAULT

        # Some constructors expect extra non-parameter items, e.g. vocab: Vocabulary.
        # We check the provided `extras` for these and just use them if they exist.
        if name in extras:
            kwargs[name] = extras[name]
        # Next case is when argument should be loaded from pretrained archive.
        elif name in params and isinstance(params.get(name), Params) and "_pretrained" in params.get(name):
            load_module_params = params.pop(name).pop("_pretrained")
            archive_file = load_module_params.pop("archive_file")
            module_path = load_module_params.pop("module_path")
            freeze = load_module_params.pop("freeze", True)
            archive = load_archive(archive_file)
            kwargs[name] = archive.extract_module(module_path, freeze) # pylint: disable=no-member
            if not isinstance(kwargs[name], annotation):
                raise ConfigurationError(f"The module from model at {archive_file} at path {module_path} "
                                         f"was expected of type {annotation} but is of type {type(kwargs[name])}")
        # # The next case is when the parameter type is itself constructible from_params.
        elif hasattr(annotation, 'from_params'):
            if name in params:
                # Our params have an entry for this, so we use that.
                subparams = params.pop(name)
                subextras = create_extras(annotation, extras)

                # In some cases we allow a string instead of a param dict, so
                # we need to handle that case separately.
                if isinstance(subparams, str):
                    kwargs[name] = annotation.by_name(subparams)()
                else:
                    kwargs[name] = annotation.from_params(params=subparams, **subextras)
            elif not optional:
                # Not optional and not supplied, that's an error!
                raise ConfigurationError(f"expected key {name} for {cls.__name__}")
            else:
                kwargs[name] = default

        # If the parameter type is a Python primitive, just pop it off
        # using the correct casting pop_xyz operation.
        elif annotation == str:
            kwargs[name] = (params.pop(name, default)
                            if optional
                            else params.pop(name))
        elif annotation == int:
            kwargs[name] = (params.pop_int(name, default)
                            if optional
                            else params.pop_int(name))
        elif annotation == bool:
            kwargs[name] = (params.pop_bool(name, default)
                            if optional
                            else params.pop_bool(name))
        elif annotation == float:
            kwargs[name] = (params.pop_float(name, default)
                            if optional
                            else params.pop_float(name))

        # This is special logic for handling types like Dict[str, TokenIndexer],
        # List[TokenIndexer], Tuple[TokenIndexer, Tokenizer], and Set[TokenIndexer],
        # which it creates by instantiating each value from_params and returning the resulting structure.
        elif origin in (Dict, dict) and len(args) == 2 and hasattr(args[-1], 'from_params'):
            value_cls = annotation.__args__[-1]

            value_dict = {}

            for key, value_params in params.pop(name, Params({})).items():
                subextras = create_extras(value_cls, extras)
                value_dict[key] = value_cls.from_params(params=value_params, **subextras)

            kwargs[name] = value_dict

        elif origin in (List, list) and len(args) == 1 and hasattr(args[0], 'from_params'):
            value_cls = annotation.__args__[0]

            value_list = []

            for value_params in params.pop(name, Params({})):
                subextras = create_extras(value_cls, extras)
                value_list.append(value_cls.from_params(params=value_params, **subextras))

            kwargs[name] = value_list

        elif origin in (Tuple, tuple) and all(hasattr(arg, 'from_params') for arg in args):
            value_list = []

            for value_cls, value_params in zip(annotation.__args__, params.pop(name, Params({}))):
                subextras = create_extras(value_cls, extras)
                value_list.append(value_cls.from_params(params=value_params, **subextras))

            kwargs[name] = tuple(value_list)

        elif origin in (Set, set) and len(args) == 1 and hasattr(args[0], 'from_params'):
            value_cls = annotation.__args__[0]

            value_set = set()

            for value_params in params.pop(name, Params({})):
                subextras = create_extras(value_cls, extras)
                value_set.add(value_cls.from_params(params=value_params, **subextras))

            kwargs[name] = value_set

        else:
            # Pass it on as is and hope for the best.   ¯\_(ツ)_/¯
            if optional:
                kwargs[name] = params.pop(name, default)
            else:
                kwargs[name] = params.pop(name)

    params.assert_empty(cls.__name__)
    return kwargs
```

**Comment by matt-gardner** [comment]
@codedecde, I just merged #2734, which made some changes that may have resolved this issue.  Can you try to reproduce your issue on the latest master, and if you succeed, give us the code you used and a stack trace?

**Comment by codedecde** [comment]
#2734 would fix the case for FeedForward, because now FeedForward inherits from FromParams, whose from_params method takes extras by default. However any class with a custom from_params would still cause the issue. For example consider
```python
from typing import List
from allennlp.common import Registrable
from allennlp.models import Model
from allennlp.modules import FeedForward
from allennlp.common.params import Params
from allennlp.data.vocabulary import Vocabulary

import torch.nn as nn


class CustomClass(nn.Module, Registrable):
    def __init__(self, input_dim: int):
        super(CustomClass, self).__init__()
        self.input_dim = input_dim

    @classmethod
    def from_params(cls, params: Params):
        """Any custom from params that doesn't accept
        extras will have the issue
        """
        input_dim = params.pop_int("input_dim")
        params.assert_empty(cls.__name__)
        return cls(input_dim)


@Model.register('tmp_model')
class TmpModel(Model):
    def __init__(self, vocab: Vocabulary, custom_class_list: List[CustomClass]):
        super(TmpModel, self).__init__(vocab=vocab)
        self.custom_class_list = nn.ModuleList(custom_class_list)


if __name__ == "__main__":
    params = Params({
        "type": "tmp_model",
        "custom_class_list": [{"input_dim": 512}]
    })

    vocab = Vocabulary()
    model = Model.from_params(vocab=vocab, params=params)
```
Would still throw the following error
```
Traceback (most recent call last):
  File "bug_repro.py", line 40, in <module>
    model = Model.from_params(vocab=vocab, params=params)
  File "/allennlp/allennlp/common/from_params.py", line 328, in from_params
    return subclass.from_params(params=params, **extras)
  File "/allennlp/allennlp/common/from_params.py", line 351, in from_params
    kwargs = create_kwargs(cls, params, **extras)
  File "/allennlp/allennlp/common/from_params.py", line 114, in create_kwargs
    kwargs[name] = construct_arg(cls, name, annotation, param.default, params, **extras)
  File "/allennlp/allennlp/common/from_params.py", line 224, in construct_arg
    value_list.append(value_cls.from_params(params=value_params, **extras))
TypeError: from_params() got an unexpected keyword argument 'vocab'
```
The potential fix could be similar to the one mentioned before 
```python
def create_extras(cls: Type[T], extras: Dict[str, Any]) -> Dict[str, Any]:
    """
    Given a dictionary of extra arguments, returns a dictionary of
    kwargs that actually are a part of the signature of the cls.from_params
    method.
    """
    subextras: Dict[str, Any] = {}
    if takes_arg(cls.from_params, "extras"):
        # If annotation.params accepts **extras, we need to pass them all along.
        # For example, `BasicTextFieldEmbedder.from_params` requires a Vocabulary
        # object, but `TextFieldEmbedder.from_params` does not.
        subextras = extras
    else:
        # Otherwise, only supply the ones that are actual args; any additional ones
        # will cause a TypeError.
        subextras = {k: v for k, v in extras.items()
                     if takes_arg(cls.from_params, k)}
    return subextras


def construct_arg(cls: Type[T], # pylint: disable=inconsistent-return-statements,too-many-return-statements
                  param_name: str,
                  annotation: Type,
                  default: Any,
                  params: Params,
                  **extras) -> Any:
    """
    Does the work of actually constructing an individual argument for :func:`create_kwargs`.

    Here we're in the inner loop of iterating over the parameters to a particular constructor,
    trying to construct just one of them.  The information we get for that parameter is its name,
    its type annotation, and its default value; we also get the full set of ``Params`` for
    constructing the object (which we may mutate), and any ``extras`` that the constructor might
    need.

    We take the type annotation and default value here separately, instead of using an
    ``inspect.Parameter`` object directly, so that we can handle ``Union`` types using recursion on
    this method, trying the different annotation types in the union in turn.
    """
    from allennlp.models.archival import load_archive  # import here to avoid circular imports

    # We used `param_name` as the method argument to avoid conflicts with 'name' being a key in
    # `extras`, which isn't _that_ unlikely.  Now that we are inside the method, we can switch back
    # to using `name`.
    name = param_name
    origin = getattr(annotation, '__origin__', None)
    args = getattr(annotation, '__args__', [])

    # The parameter is optional if its default value is not the "no default" sentinel.
    optional = default != _NO_DEFAULT

    # Some constructors expect extra non-parameter items, e.g. vocab: Vocabulary.
    # We check the provided `extras` for these and just use them if they exist.
    if name in extras:
        return extras[name]
    # Next case is when argument should be loaded from pretrained archive.
    elif name in params and isinstance(params.get(name), Params) and "_pretrained" in params.get(name):
        load_module_params = params.pop(name).pop("_pretrained")
        archive_file = load_module_params.pop("archive_file")
        module_path = load_module_params.pop("module_path")
        freeze = load_module_params.pop("freeze", True)
        archive = load_archive(archive_file)
        result = archive.extract_module(module_path, freeze) # pylint: disable=no-member
        if not isinstance(result, annotation):
            raise ConfigurationError(f"The module from model at {archive_file} at path {module_path} "
                                     f"was expected of type {annotation} but is of type {type(result)}")
        return result
    # The next case is when the parameter type is itself constructible from_params.
    elif hasattr(annotation, 'from_params'):
        if name in params:
            # Our params have an entry for this, so we use that.
            subparams = params.pop(name)

            subextras = create_extras(annotation, extras)

            # In some cases we allow a string instead of a param dict, so
            # we need to handle that case separately.
            if isinstance(subparams, str):
                return annotation.by_name(subparams)()
            else:
                return annotation.from_params(params=subparams, **subextras)
        elif not optional:
            # Not optional and not supplied, that's an error!
            raise ConfigurationError(f"expected key {name} for {cls.__name__}")
        else:
            return default

    # If the parameter type is a Python primitive, just pop it off
    # using the correct casting pop_xyz operation.
    elif annotation == str:
        return params.pop(name, default) if optional else params.pop(name)
    elif annotation == int:
        return params.pop_int(name, default) if optional else params.pop_int(name)
    elif annotation == bool:
        return params.pop_bool(name, default) if optional else params.pop_bool(name)
    elif annotation == float:
        return params.pop_float(name, default) if optional else params.pop_float(name)

    # This is special logic for handling types like Dict[str, TokenIndexer],
    # List[TokenIndexer], Tuple[TokenIndexer, Tokenizer], and Set[TokenIndexer],
    # which it creates by instantiating each value from_params and returning the resulting structure.
    elif origin in (Dict, dict) and len(args) == 2 and hasattr(args[-1], 'from_params'):
        value_cls = annotation.__args__[-1]

        value_dict = {}

        for key, value_params in params.pop(name, Params({})).items():
            subextras = create_extras(value_cls, extras)
            value_dict[key] = value_cls.from_params(params=value_params, **subextras)

        return value_dict

    elif origin in (List, list) and len(args) == 1 and hasattr(args[0], 'from_params'):
        value_cls = annotation.__args__[0]

        value_list = []

        for value_params in params.pop(name, Params({})):
            subextras = create_extras(value_cls, extras)
            value_list.append(value_cls.from_params(params=value_params, **subextras))

        return value_list

    elif origin in (Tuple, tuple) and all(hasattr(arg, 'from_params') for arg in args):
        value_list = []

        for value_cls, value_params in zip(annotation.__args__, params.pop(name, Params({}))):
            subextras = create_extras(value_cls, extras)
            value_list.append(value_cls.from_params(params=value_params, **subextras))

        return tuple(value_list)

    elif origin in (Set, set) and len(args) == 1 and hasattr(args[0], 'from_params'):
        value_cls = annotation.__args__[0]

        value_set = set()

        for value_params in params.pop(name, Params({})):
            subextras = create_extras(value_cls, extras)
            value_set.add(value_cls.from_params(params=value_params, **subextras))

        return value_set

    elif origin == Union:
        # Storing this so we can recover it later if we need to.
        param_value = params.get(name, Params({}))
        if isinstance(param_value, Params):
            param_value = param_value.duplicate()

        # We'll try each of the given types in the union sequentially, returning the first one that
        # succeeds.
        for arg in args:
            try:
                return construct_arg(cls, name, arg, default, params, **extras)
            except (ValueError, TypeError, ConfigurationError, AttributeError):
                # Our attempt to construct the argument may have popped `params[name]`, so we
                # restore it here.
                params[name] = param_value
                if isinstance(param_value, Params):
                    param_value = param_value.duplicate()
                continue

        # If none of them succeeded, we crash.
        raise ConfigurationError(f"Failed to construct argument {name} with type {annotation}")
    else:
        # Pass it on as is and hope for the best.   ¯\_(ツ)_/¯
        if optional:
            return params.pop(name, default)
        else:
            return params.pop(name)
```

Would be happy to create a PR : )

**Comment by matt-gardner** [comment]
Yep, thanks, having now recently touched this logic myself, I see the problem and agree with your solution.  PR welcome!
